### DISCUSSION TITLE
resolveComponent can only be used in render() or setup()

### DISCUSSION THREAD

**Discussion by joshstrange** [quasarframework/quasar#16084] [General - Components / Directives / etc]
I've spent 8+ hours on this over the last 2 days and I wanted to record my final findings here to hopefully help at least 1 other person.

I had been getting this warning which resulted in a white screen being shown instead of my app:
 
    resolveComponent can only be used in render() or setup()

This wasn't the only warning I saw, I don't have the other ones now and only the one above is happening now. I also got a weird error about trying to access `@fs/Users/myname/path/to/code/node_modules/.pnpm/....` in the browser console (and in the dev server output) a few times. All these errors were "sticky" (once it started it didn't start until "something" changed, keep reading).

This issue was incredibly inconsistent and would sometimes go away and my app would work just fine for a random amount of time. I found no rhyme or reason to what caused it to fix itself. Sometimes on the same dev server "run" it would work, then stop, then work again with no real code changes and, when I'd revert the 1-line change I had made, it would not flip again from working->not-working or not-working->working.

A few other things I tried that had zero effect. I mean sometimes it would "fix" my site but it was never reproducible. I tried every combination and in different orders:

* Delete all node_modules + reinstall
* Delete `.quasar`
* Delete `node_modules/.q-cache`
* Restart dev server (ctrl+c, then re-reun)
* Restart dev server (by editing `quasar.config.js`)
* Upgrading all packages to newest supported versions
* Commenting out huge swaths of my codebase
* Probably other crazy things

The ONLY thing that reliably "fixed" it was commenting out the `<router-view></router-view>` in my `App.vue` and just putting static text there. That would work, nothing else would work and I tried removing all my routes and using a single clean component in a route to no avail.

I cannot stress enough, the inconsistent nature of this issue was infuriating. It would start working and I thought I had fixed it and I'd start working on the real work I need to do then it happen again a few minutes or more later.

A huge number of my google searches pointed at the fact that I MUST be running more than 1 instance of Vue. That sort of made sense, I understood the concept but I was pretty sure I wasn't. Nevertheless I tried every solution posted:

* `viteConf.optimizeDeps.exclude`
* `viteConf.rollupOptions.external`
* `viteConf.resolve.dedupe`
* viteExternalsPlugin
* More that I've forgotten

I tried multiple permutations of the above. Sometimes they would "fix" it, then a few hot reloads later it would be broken and undoing everything I had done didn't fix it. I think something about restarting the dev server made this work randomly. Again, to reiterate, restarting the dev server did not fix it most of the time.

I finally spun up a fresh Quasar project to try to reproduce but didn't get far before hitting an odd error after copying over my `devServer` config. It was failing when trying to fetch something that from my main project and the fresh Quasar install should have no idea about. I didn't even copy over the `devServer` config to test it, I was just copying my full `quasar.config.js` file to see if it was something in there that was causing the issue, turns out there was.

Here is what my `devServer` looked like:

```
		devServer: {
			origin: 'https://local.MYDOMAIN.com:443',
			port: 9500,
			open: false, // opens browser window automatically
			vueDevtools: true,
			hmr: {
				host: 'ws.MYDOMAIN.com',
				protocol: 'wss',
				clientPort: 443,
				// port: 443,
			}
		},
```

I had 2 CloudFront Tunnels setup like so (yml config file that I ran with `cloudflared tunnel --config ~/.cloudflared/NAME.yml run NAME`):

```
tunnel: REMOVED
credentials-file: /Users/josh/.cloudflared/REMOVED.json


ingress:
  - hostname: local.MYDOMAIN.com
    service: http://MYLOCALIP:9500
  - hostname: ws.MYDOMAIN.com
    service: ws://MYLOCALIP:9500
  - service: http_status:404

```

I have no clue why going through a Cloudflare tunnel would cause such a weird issue. My first thought is caching but I have caching disabled when I have dev tools open (I always had dev tools open during this) and I was able to reproduce the issue in more than 1 browser and in incognito mode. 

I am completely stumped but after switching to the default config:

```
		devServer: {
			open: true // opens browser window automatically
		},

```

My site reliably worked and has worked for well over an hour (about 50min longer than the longest it has worked in the past).

I developed with this config for weeks in the past, it's how I developed this site for a long time. I haven't changed anything that I can think of, I upgraded Cloudflare's tunnel tool months ago but this setup has worked since then. 

Here are the keywords I was searching for along with the error message:

pnpm
vite
vue
quasar
vue-router

Here is my Quasar info:

```
Operating System - Darwin(22.4.0) - darwin/arm64
NodeJs - 16.19.0

Global packages
  NPM - 8.19.3
  yarn - Not installed
  @quasar/cli - 2.2.1
  @quasar/icongenie - 2.5.4
  cordova - Not installed

Important local packages
  quasar - 2.12.2 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app-vite - 1.4.3 -- Quasar Framework App CLI with Vite
  @quasar/extras - 1.16.5 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.3.4 -- The progressive JavaScript framework for building modern web UI.
  vue-router - 4.2.4
  pinia - 2.1.4 -- Intuitive, type safe and flexible Store for Vue
  vuex - Not installed
  vite - Not installed
  eslint - 8.22.0 -- An AST-based pattern checker for JavaScript.
  electron - Not installed
  electron-packager - Not installed
  electron-builder - Not installed
  register-service-worker - Not installed
  @capacitor/core - 4.7.0 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/cli - 4.7.0 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/android - 4.7.0 -- Capacitor: Cross-platform apps with JavaScript and the web
  @capacitor/ios - 4.7.0 -- Capacitor: Cross-platform apps with JavaScript and the web

Quasar App Extensions
  *None installed*
```

I really hope I save someone some time with this.

**Comment by zemf4you** [comment]
Same situation - I use cloudflared and I am facing with this issue too. Terrible bug, very annoying and random. But my app is dockerized, so `open: true` doesn't help.
Disappointing

**Reply by zemf4you** [reply]
A few moments to think about:
1) it may be a matter of Vue duplication. Try resolve.dedupe for the Vite config.
2) you may be calling useQuasar not in the component's setup/render. This can randomly cause an error, although it is consistently reproduced on a minimal example.
3) there is a suspicion of routerview from vue-router or what is happening using it (for example, dynamic import of components).
This problem definitely has a random race condition, but its causes no longer seem random.

**Comment by Valgri** [comment]
same problem, but with an external library. I have tried everything described here, nothing helps. Who can help? 
![2024-02-05_17-26](https://github.com/quasarframework/quasar/assets/28981859/3f24928d-d0ae-47b2-b6df-74a3741136e1)

**Comment by zemf4you** [comment]
Well, since this problem has signs of a race condition, you can try the following:
1) enable open: true. This really causes the index page to warm up, as the vite documentation says.
2) server.warmup.
3) optimizeDeps.
4) disabling sourcemaps.
5) just buy more powerful pc.
